### ISSUE TITLE
Combine Multi-datasets with batch effect correction 

### ISSUE THREAD

**Issue by Two-J** [satijalab/seurat#3505]
Hello, 
I tried merging multi-datasets with batch effect correction, referring to vignettes of Satijalab/Seurat/Integration and Label Transfer.
As follows, my workflow was ...
Combining data -> Normalize Data -> Find variable features -> Batch correction (FindIntegrationAnchors) -> Scale -> Run PCA/UMAP -> Quality Controls -> Normalize and Find variable features again (but with this step, the batch correction seemed to be removed) -> Scale, Run PCA/UMAP -> Subset data (only immune cells) -> Normalize and Find variable features again, scale, run PCA/UMAP -> (since I confirmed the corrected batch effect was removed) Batch correction again. -> Scale, Run PCA/UMAP, clustering ->  plotting.

Here below is all the codes I used. I think the final plot (after final batch correction) is merged well, nevertheless, I am really concerned if these codes are wrong :( I desperately need expert opinions! Thank you!

```r
####Combine Data####
A_combined<- merge(x=A1, y = c(A2,A3,A4,A5,A6), add.cell.ids = c("A1","A2","A3","A4","A5","A6"), project = "A")
A_combined
table(A_combined$orig.ident)
####Batch-effect Correction with Seurat CCA Method, FindIntegrationAnchors####
A_combined.list <- SplitObject(A_combined,split.by = "orig.ident")
A_combined.list <- A_combined.list[c("A1","A2","A3","A4","A5","A6")]
for (i in 1:length(A_combined.list)) {
  A_combined.list[[i]] <- NormalizeData(A_combined.list[[i]], verbose = FALSE)
  A_combined.list[[i]] <- FindVariableFeatures(A_combined.list[[i]], selection.method = "vst", 
                                                   nfeatures = 2000, verbose = FALSE)
}
reference.list <- A_combined.list[c("A1","A2","A3","A4","A5","A6")]
A.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30)
A_integrated <- IntegrateData(anchorset = A.anchors, dims = 1:30)
DefaultAssay(A_integrated) <- "integrated"
A_integrated <- ScaleData(A_integrated, verbose = FALSE)
A_integrated <- RunPCA(A_integrated, npcs = 30, verbose = FALSE)
A_integrated <- RunUMAP(A_integrated, reduction = "pca", dims = 1:30)
DimPlot(A_integrated, reduction = "umap", group.by = "orig.ident")
DimPlot(A_integrated, reduction = "umap", split.by = "orig.ident")

DefaultAssay(A_integrated) <- "RNA"
A_integrated[["percent.mt"]] <- PercentageFeatureSet(A_integrated, pattern = "^MT-", assay = 'RNA')
A_integrated[["percent.rps"]] <- PercentageFeatureSet(A_integrated, pattern = "^RPS-", assay = 'RNA')
VlnPlot(A_integrated, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rps"), ncol = 3)
A_integrated_afterQC <- subset(A_integrated, subset = nFeature_RNA > 100 & nFeature_RNA < 7500 & percent.mt < 15)
A_integrated_afterQC <- NormalizeData(A_integrated_afterQC, normalization.method = "LogNormalize", scale.factor = 10000)
A_integrated_afterQC <- FindVariableFeatures(A_integrated_afterQC, selection.method = "vst", nfeatures = 2000)
top30 <- head(VariableFeatures(A_integrated_afterQC), 30)
plot1 <- VariableFeaturePlot(A_integrated_afterQC)
plot2 <- LabelPoints(plot = plot1, points = top30, repel = TRUE)
plot1 + plot2

all.genes <- rownames(A_integrated_afterQC)
A_integrated_afterQC <- ScaleData(A_integrated_afterQC, features = all.genes)
A_integrated_afterQC <- RunPCA(A_integrated_afterQC, features = VariableFeatures(object = A_integrated_afterQC))
print(A_integrated_afterQC[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(A_integrated_afterQC, dims = 1:2, reduction = "pca")
DimPlot(A_integrated_afterQC, reduction = "pca")
DimHeatmap(A_integrated_afterQC, dims = 1, cells = 500, balanced = TRUE)
ElbowPlot(A_integrated_afterQC)

A_integrated_afterQC <- FindNeighbors(A_integrated_afterQC, dims = 1:40)
A_integrated_afterQC <- FindClusters(A_integrated_afterQC, resolution = 0.7)
A_integrated_afterQC <- RunUMAP(A_integrated_afterQC, dims = 1:40)
DimPlot(A_integrated_afterQC, reduction = "umap", pt.size = 1) ##-> problem 1: here, the plot is returned to the state before batch correction##


A_integrated_afterQC_immune<-SubsetData(object=A_integrated_afterQC, ident.use =c(0,9,11,23,8), do.clean=TRUE, do.scale=TRUE)
A_integrated_afterQC_immune <- NormalizeData(A_integrated_afterQC_immune, normalization.method = "LogNormalize", scale.factor = 10000)
A_integrated_afterQC_immune <- FindVariableFeatures(A_integrated_afterQC_immune, selection.method = "vst", nfeatures = 2000)
top30 <- head(VariableFeatures(A_integrated_afterQC_immune), 30)
plot1 <- VariableFeaturePlot(A_integrated_afterQC_immune)
plot2 <- LabelPoints(plot = plot1, points = top30, repel = TRUE)
plot1 + plot2
all.genes <- rownames(A_integrated_afterQC_immune)
A_integrated_afterQC_immune <- ScaleData(A_integrated_afterQC_immune, features = all.genes)
A_integrated_afterQC_immune <- RunPCA(A_integrated_afterQC_immune, features = VariableFeatures(object = A_integrated_afterQC_immune))
print(A_integrated_afterQC_immune[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(A_integrated_afterQC_immune, dims = 1:2, reduction = "pca")
DimPlot(A_integrated_afterQC_immune, reduction = "pca")
DimHeatmap(A_integrated_afterQC_immune, dims = 1, cells = 500, balanced = TRUE)
ElbowPlot(A_integrated_afterQC_immune)

A_integrated_afterQC_immune <- FindNeighbors(A_integrated_afterQC_immune, dims = 1:20)
A_integrated_afterQC_immune <- FindClusters(A_integrated_afterQC_immune, resolution = 0.6)
A_integrated_afterQC_immune <- RunUMAP(A_integrated_afterQC_immune, dims = 1:20)
DimPlot(A_integrated_afterQC_immune, reduction = "umap", pt.size = 1, label = TRUE) ##-> Likewise, this plot is shown without batch correction

##so I performed bath correction again here##
A_integrated_afterQC_immune.list <- SplitObject(A_integrated_afterQC_immune,split.by = "orig.ident")
A_integrated_afterQC_immune.list <- A_integrated_afterQC_immune.list[c("GC1","GC2","GC3","GC4","GC5","GC6")]
for (i in 1:length(A_integrated_afterQC_immune.list)) {
  A_integrated_afterQC_immune.list[[i]] <- NormalizeData(A_integrated_afterQC_immune.list[[i]], verbose = FALSE)
  A_integrated_afterQC_immune.list[[i]] <- FindVariableFeatures(A_integrated_afterQC_immune.list[[i]], selection.method = "vst", 
                                                             nfeatures = 2000, verbose = FALSE)
}
reference.list <- A_integrated_afterQC_immune.list[c("GC1","GC2","GC3","GC4","GC5","GC6")]
A_immune.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30)
A_integrated_immune <- IntegrateData(anchorset = A_immune.anchors, dims = 1:30)
DefaultAssay(A_integrated_immune) <- "integrated"
A_integrated_immune <- ScaleData(A_integrated_immune, verbose = FALSE)
A_integrated_immune <- RunPCA(A_integrated_immune, npcs = 30, verbose = FALSE)
A_integrated_immune <- RunUMAP(A_integrated_immune, reduction = "pca", dims = 1:30)
A_integrated_immune <- FindNeighbors(A_integrated_immune, dims = 1:20)
A_integrated_immune <- FindClusters(A_integrated_immune, resolution = 0.02)
A_integrated_immune <- RunUMAP(A_integrated_immune, dims = 1:20)
DimPlot(A_integrated_immune, reduction = "umap", pt.size = 1, label = FALSE)
```

**Comment by Two-J** [comment]
oops! my mistake!
please see "GC1","GC2","GC3","GC4","GC5","GC6" = "A1","A2","A3","A4","A5","A6"

Thank you so much, 
Two-J

**Comment by timoast** [comment]
We unfortunately don't have the bandwidth to provide detailed code review, but I will note a few points:

1. There's no point in merging and the splitting the objects (first few lines). This won't affect the results but is unnecessary.
2. I'd suggest doing QC and filtering cells on each object before running the integration.
3. Running NormalizeData on the integrated assay will overwrite the integration results.

**Comment by McStber** [comment]
Dear @timoast , dear community,

One general question connected to merging questions from @Beyondvisible :

I have an immune cell dataset containing different treatments (more than one group thats hwy I did Integration). 
I subsetted the lymphoctes from the myeloids.
This successfully led to better resolved lymphocytes clusters :) 

Now I want to merge my "new/better clustered lymphocytes" with my original dataset.

Is it correct if I am 

1. (in my original dataset) excluding the lymphocytes 
2. merge my new lymphocytes with my old dataset (which contains now only myeloids)
3. Redo normalizing, FindVariableFeatures, Scaling, RunPCA, RunUMAP (same dims as before), FindNeighbors (same dims as before), FindClusters?

Will the again normalization affect the gene expression?
Is there another way to use the cell IDs/barcodes to trace or identify back my new lymphocytes in my orginal dataset and redo a clustering?

I would be very happy to get your feedback and your opinion.
Many thanks in advance.

Best,
McStber
